go/arrow/util/protobuf_reflect_test.go

```diff
@@ -0,0 +1,247 @@
+	want = `schema:
+  fields: 21
+    - string: type=utf8, nullable
+    - int32: type=int32, nullable
+    - int64: type=int64, nullable
+    - sint32: type=int32, nullable
+    - sin64: type=int64, nullable
+    - uint32: type=uint32, nullable
+    - uint64: type=uint64, nullable
+    - fixed32: type=uint32, nullable
+    - fixed64: type=uint64, nullable
+    - sfixed32: type=int32, nullable
+    - bool: type=bool, nullable
+    - bytes: type=binary, nullable
+    - double: type=float64, nullable
+    - enum: type=dictionary<values=utf8, indices=int32, ordered=false>, nullable
+    - message: type=struct<field1: utf8>, nullable
+    - oneof: type=dense_union<oneofstring: type=utf8, nullable=0, oneofmessage: type=struct<field1: utf8>, nullable=1>, nullable
+    - any: type=struct<field1: utf8>, nullable
+    - simple_map: type=map<int32, utf8, items_nullable>, nullable
+    - complex_map: type=map<utf8, struct<field1: utf8>, items_nullable>, nullable
+    - simple_list: type=list<item: utf8, nullable>, nullable
+    - complex_list: type=list<item: struct<field1: utf8>, nullable>, nullable`
+
+	require.Equal(t, want, got, "got: %s\nwant: %s", got, want)
+
+	excludeComplex := func(pfr *protobufFieldReflection) bool {
+		return pfr.isMap() || pfr.isList() || pfr.isStruct()
+	}
+
+	got = NewSuperMessage(&msg, WithExclusionPolicy(excludeComplex)).Schema().String()
+	want = `schema:
+  fields: 15
+    - string: type=utf8, nullable
+    - int32: type=int32, nullable
+    - int64: type=int64, nullable
+    - sint32: type=int32, nullable
+    - sin64: type=int64, nullable
+    - uint32: type=uint32, nullable
+    - uint64: type=uint64, nullable
+    - fixed32: type=uint32, nullable
+    - fixed64: type=uint64, nullable
+    - sfixed32: type=int32, nullable
+    - bool: type=bool, nullable
+    - bytes: type=binary, nullable
+    - double: type=float64, nullable
+    - enum: type=dictionary<values=utf8, indices=int32, ordered=false>, nullable
+    - oneofstring: type=utf8, nullable`
+
+	require.Equal(t, want, got, "got: %s\nwant: %s", got, want)
+
+	got = NewSuperMessage(
+		&msg,
+		WithExclusionPolicy(excludeComplex),
+		WithFieldNameFormatter(xstrings.ToCamelCase),
+	).Schema().String()
+	want = `schema:
+  fields: 15
+    - String: type=utf8, nullable
+    - Int32: type=int32, nullable
+    - Int64: type=int64, nullable
+    - Sint32: type=int32, nullable
+    - Sin64: type=int64, nullable
+    - Uint32: type=uint32, nullable
+    - Uint64: type=uint64, nullable
+    - Fixed32: type=uint32, nullable
+    - Fixed64: type=uint64, nullable
+    - Sfixed32: type=int32, nullable
+    - Bool: type=bool, nullable
+    - Bytes: type=binary, nullable
+    - Double: type=float64, nullable
+    - Enum: type=dictionary<values=utf8, indices=int32, ordered=false>, nullable
+    - Oneofstring: type=utf8, nullable`
+
+	require.Equal(t, want, got, "got: %s\nwant: %s", got, want)
+}
+
+func TestRecordFromProtobuf(t *testing.T) {
```

Comment:
can we do a round trip test?

Reply:
I tried this:
```
func TestRoundTrip(t *testing.T) {
	got := SetupTest()
	b, _ := NewProtobufMessageReflection(&got).Record(nil).MarshalJSON()
	fmt.Println(string(b))
	want := util_message.AllTheTypes{}
	err := jsonpb.Unmarshal(bytes.NewReader(b), &want)

	require.NoError(t, err)
	require.Equal(t, got, want, "got: %s\nwant: %s", got, want)
}
```

but the arrow -> json, wraps the map in a list so it's a bit tricky. 
I could implement ProtobufMessageReflection -> struct, but that will take a bit of work.

Reply:
It's fine for RoundTrip to be a follow-up PR, no worries